genesyscloud/integration_action/resource_genesyscloud_integration_action.go

```diff
@@ -0,0 +1,275 @@
+	if diagErr != nil {
+		return diagErr
+	}
+
+	diagErr = gcloud.RetryWhen(gcloud.IsStatus400, func() (*platformclientv2.APIResponse, diag.Diagnostics) {
+		action, resp, err := iap.createIntegrationAction(ctx, &IntegrationAction{
+			Name:          &name,
+			Category:      &category,
+			IntegrationId: &integrationId,
+			Secure:        &secure,
+			Contract:      actionContract,
+			Config:        buildSdkActionConfig(d),
+		})
+		if err != nil {
+			return resp, diag.Errorf("Failed to create integration action %s: %s", name, err)
+		}
+		d.SetId(*action.Id)
+
+		log.Printf("Created integration action %s %s", name, *action.Id)
+		return resp, nil
+	})
+	if diagErr != nil {
+		return diagErr
+	}
+
+	return readIntegrationAction(ctx, d, meta)
+}
+
+// readIntegrationAction is used by the integration action resource to read an action from genesys cloud.
+func readIntegrationAction(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	sdkConfig := meta.(*gcloud.ProviderMeta).ClientConfig
+	iap := getIntegrationActionsProxy(sdkConfig)
+
+	log.Printf("Reading integration action %s", d.Id())
+
+	return gcloud.WithRetriesForRead(ctx, d, func() *retry.RetryError {
+		action, resp, err := iap.getIntegrationActionById(ctx, d.Id())
+		if err != nil {
+			if gcloud.IsStatus404(resp) {
+				return retry.RetryableError(fmt.Errorf("failed to read integration action %s: %s", d.Id(), err))
+			}
+			return retry.NonRetryableError(fmt.Errorf("failed to read integration action %s: %s", d.Id(), err))
+		}
+
+		// Retrieve config request/response templates
+		reqTemp, resp, err := iap.getIntegrationActionTemplate(ctx, d.Id(), reqTemplateFileName)
+		if err != nil {
+			if gcloud.IsStatus404(resp) {
+				d.SetId("")
+				return nil
+			}
+			return retry.NonRetryableError(fmt.Errorf("failed to read request template for integration action %s: %s", d.Id(), err))
+		}
+
+		successTemp, resp, err := iap.getIntegrationActionTemplate(ctx, d.Id(), successTemplateFileName)
+		if err != nil {
+			if gcloud.IsStatus404(resp) {
+				d.SetId("")
+				return nil
+			}
+			return retry.NonRetryableError(fmt.Errorf("failed to read success template for integration action %s: %s", d.Id(), err))
+		}
+
+		cc := consistency_checker.NewConsistencyCheck(ctx, d, meta, ResourceIntegrationAction())
+		if action.Name != nil {
+			d.Set("name", *action.Name)
+		} else {
+			d.Set("name", nil)
+		}
+
+		if action.Category != nil {
+			d.Set("category", *action.Category)
+		} else {
+			d.Set("category", nil)
+		}
+
+		if action.IntegrationId != nil {
+			d.Set("integration_id", *action.IntegrationId)
+		} else {
+			d.Set("integration_id", nil)
+		}
+
+		if action.Secure != nil {
+			d.Set("secure", *action.Secure)
+		} else {
+			d.Set("secure", nil)
+		}
+
+		if action.Config.TimeoutSeconds != nil {
+			d.Set("config_timeout_seconds", *action.Config.TimeoutSeconds)
+		} else {
+			d.Set("config_timeout_seconds", nil)
+		}
```

Comment:
For these operations, I recommend using the function SetNillableValue in the resourcedata package. It uses generics to perform all the `if T is nil { d.Set(key, *T) } else { d.Set(key, nil) }` logic in one function. 

Here is an example implementation: [https://github.com/MyPureCloud/terraform-provider-genesyscloud/blob/main/genesyscloud/routing_sms_addresses/resource_genesyscloud_routing_sms_addresses.go#L109](https://github.com/MyPureCloud/terraform-provider-genesyscloud/blob/main/genesyscloud/routing_sms_addresses/resource_genesyscloud_routing_sms_addresses.go#L109)

Reply:
Used resourcedata util functions where applicable